### PR TITLE
Update poly dec

### PR DESCRIPTION
**Pull request by JacquesOlivierLachaud**
# PR Description

Updates to PolygonalCalculus : sign, global matrices, dirichlet conditions, Poisson examples

# Checklist

- [x] Unit-test of your feature with [Catch](http://dgtal.org/doc/stable/moduleCatch.html).
- [x] Doxygen documentation of the code completed (classes, methods, types, members...)
- [x] Documentation module page added or updated.
- [x] New entry in the [ChangeLog.md](https://github.com/DGtal-team/DGtal/blob/master/ChangeLog.md) added.
- [x] No warning raised in Debug mode.
- [x] All continuous integration tests pass (Github Actions & appveyor)

### CODE REVIEW

**Comment by JacquesOlivierLachaud** [comment]
I did hesitate on adding global matrices for sharp/flat, but perhaps this is not the spirit of this dec.
Also, did I add the half-sphere example for geodesics on surfaces with boundary ?

File: src/DGtal/dec/PolygonalCalculus.h [hunk 1]
```diff
@@ -435,7 +447,7 @@ class PolygonalCalculus
     if (checkCache(DIVERGENCE_,f))
       return myGlobalCache[DIVERGENCE_][f];
  
-    DenseMatrix op = D(f).transpose() * M(f);
+    DenseMatrix op = -1.0 * D(f).transpose() * M(f);
```
Comment (dcoeurjo) [inline src/DGtal/dec/PolygonalCalculus.h:450]: Please comment in the method description that the change in sign is just a convention and that this one does not match with @cite degoes2020discrete

BTW, this convention does not match with Gradi's (positive values on diag)... It's fine for me if it is discussed

File: src/DGtal/dec/PolygonalCalculus.h [hunk 2]
```diff
@@ -465,54 +477,57 @@ class PolygonalCalculus
     if (checkCache(L_,f))
       return myGlobalCache[L_][f];
  
-    auto Df = D(f);
-    DenseMatrix op = Df.transpose() * M(f,lambda) * Df;
+    DenseMatrix Df = D(f);
+    // Laplacian is a negative operator.
+    DenseMatrix op = -1.0 * Df.transpose() * M(f,lambda) * Df;
```
Comment (dcoeurjo) [inline src/DGtal/dec/PolygonalCalculus.h:482]: Same here, the convention should be described in the method comment (I cannot suggest edits from GitHub)

File: src/DGtal/dec/PolygonalCalculus.h [hunk 3]
```diff
@@ -465,54 +477,57 @@ class PolygonalCalculus
...
-        ++cpt;
-      }
-      
       for(auto i=0; i < nf; ++i)
         for(auto j=0; j < nf; ++j)
         {
           auto v = Lap(i,j);
           if (v!= 0.0)
-            triplets.emplace_back(Triplet(reorder[i],reorder[j],Lap(i,j)));
+            triplets.emplace_back( Triplet( vertices[ i ], vertices[ j ],
```
Comment (dcoeurjo) [inline src/DGtal/dec/PolygonalCalculus.h:526]: thx.. way simpler

File: src/DGtal/math/linalg/DirichletConditions.h [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+//////////////////////////////////////////////////////////////////////////////
+
+namespace DGtal
+{
+
+  /////////////////////////////////////////////////////////////////////////////
+  // template class DirichletConditions
+  /**
+     Description of template class 'DirichletConditions' <p> \brief Aim:
+     A helper class to solve system with Dirichlet boundary conditions.
```
Comment (dcoeurjo) [inline src/DGtal/math/linalg/DirichletConditions.h:58]:
```suggestion
     A helper class to solve a system with Dirichlet boundary conditions.
```

File: src/DGtal/math/linalg/DirichletConditions.h [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+    /// is a boundary node, `p[i]=0` otherwise.
+    /// @param u the vector giving the constrained Dirichlet values on
+    /// places such that `p[i]=1`.
+    ///
+    /// @pre `#row(A) = #col(A) = #col(p) = #col(b) = #col(u)`
+    /// @return the (smaller) form \f$ b_d \f$ to solve for
+    static
+    DenseVector dirichletVector( const SparseMatrix& A,
+                                 const DenseVector& b,
+                                 const DenseVector& p,
```
Comment (dcoeurjo) [inline src/DGtal/math/linalg/DirichletConditions.h:171]: same here.. IntegerVector instead?

File: src/DGtal/math/linalg/DirichletConditions.h [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+    ///
+    /// @param xd the solution to the smaller system \f$ A_d x_d = b_d \f$.
+    /// @param p the vector such that `p[i]=1` whenever the i-th node
+    /// is a boundary node, `p[i]=0` otherwise.
+    /// @param u the vector giving the constrained Dirichlet values on
+    /// places such that `p[i]=1`.
+    /// @return the solution \f$ x \f$ to the original system \f$ A x = b \f$,
+    static
+    DenseVector dirichletSolution( const DenseVector& xd,
+                                   const DenseVector& p,
```
Comment (dcoeurjo) [inline src/DGtal/math/linalg/DirichletConditions.h:203]: Here too

**Review by dcoeurjo — COMMENTED** [review]
Sorry for the delay, many thanks for the PR. Just couple of comments.

File: ChangeLog.md [hunk 7]
```diff
@@ -37,6 +37,10 @@
     Geodesics in Heat approach and the new differential operators on
     polygonal surfaces (digital surfaces, or any PolygonalMesh instance) (David
     Coeurjolly, [#1603](https://github.com/DGtal-team/DGtal/pull/1603)
+  - Updates to PolygonalCalculus: correct sign, fix some Eigen
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply ChangeLog.md:40]: yes :)

File: examples/polyscope-examples/dgtalCalculus-poisson.cpp [hunk 8]
```diff
@@ -47,7 +48,11 @@ typedef Shortcuts<Z3i::KSpace>         SH3;
 typedef ShortcutsGeometry<Z3i::KSpace> SHG3;
 // The following typedefs are useful
 typedef SurfaceMesh< RealPoint, RealVector >  SurfMesh;
-
+typedef PolygonalCalculus<SH3::RealPoint,SH3::RealVector> PCalculus;
+typedef PCalculus::SparseMatrix SparseMatrix;
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply examples/polyscope-examples/dgtalCalculus-poisson.cpp:52]: done

File: examples/polyscope-examples/dgtalCalculus-poisson.cpp [hunk 9]
```diff
@@ -47,7 +48,11 @@ typedef Shortcuts<Z3i::KSpace>         SH3;
...
+typedef PCalculus::Vector       Form;
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply examples/polyscope-examples/dgtalCalculus-poisson.cpp:53]: cleaned

File: src/DGtal/dec/doc/modulePolygonalCalculus.dox [hunk 10]
```diff
@@ -208,6 +208,12 @@ To solve some global PDE (e.g. Laplace/Poisson problems, see below), one can com
 
 For example, the PolygonalCalculus<SurfMesh>::globalLaplaceBeltrami() method outputs a global (sparse) Laplace-Beltrami operator which can later be used for diffusion.
 
+You can also retrieve the global mass matrix via
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/dec/doc/modulePolygonalCalculus.dox:211]: done

File: src/DGtal/dec/PolygonalCalculus.h (hunk 1 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/dec/PolygonalCalculus.h:450]: done

File: src/DGtal/math/linalg/DirichletConditions.h (hunk 4 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/math/linalg/DirichletConditions.h:58]: ok

File: src/DGtal/math/linalg/DirichletConditions.h (hunk 5 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/math/linalg/DirichletConditions.h:171]: done

**Review by dcoeurjo — APPROVED** [review]
all good, thx!

**Pull request merged**